### PR TITLE
INT-3654: Rework TCP Connection Close

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3654

Previously, single-use `TcpConnections` self-closed when their use was complete.
This is unnatural and caused issues such as INT-3722.

Remove the self-closing behavior; connection close is now (properly) the responsibility
of the client using the connection:
- Client Side:
  -- `TcpOutboundGateway` after the reply is received
  -- `TcpSendingMessageHandler` after the send, when there is no collaborating inbound adapter
  -- `TcpReceivingChannelAdapter` when it is collaborating (after receiving the reply)
- Server Side:
  -- `TcpInboundGateway` after the reply is sent
  -- `TcpReceivingChannelAdapter` after the receive, when there is no collaborating outbound adapter
  -- `TcpSendingMessageHandler` when it is colllaborating (after sending the reply)

As before, the `CachingClientConnectionFactory` always sets single use on the target factory
to force it to create new connections on demand. It is always a single-use factory itself
so the clients return the connections to the pool (via `close()`).

**Needs a migration guide entry**

### CODE REVIEW

**Comment by artembilan** [comment]
This PR requires rebase and fixes for conflicts.
Thanks

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpInboundGateway.java [hunk 1]
```diff
@@ -70,29 +72,44 @@
...
+			return false;
+		}
+		finally {
+			String connectionId = (String) message.getHeaders().get(IpHeaders.CONNECTION_ID);
+			if (connectionId != null && !isErrorMessage && this.isSingleUse) {
+				if (this.serverConnectionFactory != null) {
+					this.serverConnectionFactory.closeConnection(connectionId);
+				}
+				else {
+					this.clientConnectionFactory.closeConnection(connectionId);
```
Comment (artembilan) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpInboundGateway.java]: It doesn't relate to this PR, but looks like we really should do refactoring polishing around these `server/client` variables. I don't see so much different in usage here. 
So, the final code may be more simpler. 
Like background task :smile: 
Seems for me we haven't reached any scheduled background tasks yet...

**Comment by artembilan** [comment]
OK. Nothing to complain about!
I'm trying to fix conflicts my self and merging...

**Comment by garyrussell** [comment]
No; I found a problem while fixing the conflicts.

**Comment by artembilan** [comment]
OK. Thanks. Waiting and reviewing anything else :smile:

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpInboundGateway.java (hunk 1 above)
Comment (garyrussell) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpInboundGateway.java]: Yeah - we can probably have a single factory and a boolean to decide the type.

**Comment by garyrussell** [comment]
Rebased, pushed with fix for timing problem with nested Failover and Caching CFs.

**Comment by artembilan** [comment]
Well, since I can't test it locally, I'll merge it as is, because review doesn't show any vulnerabilities.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/ec5230a

**Pull request closed without merging**